### DISCUSSION TITLE
Factories with relationship fields are calling create() when calling make()

### DISCUSSION THREAD

**Discussion by robertoarruda** [laravel/framework#36169] [Q&A]
- Laravel Version: 8.22.0
- PHP Version: 7.4.13

### Description:
I use factories in my unit tests. My tests never touch the database, so I don't use connections or migrations on it.
In my unit tests, factories use the make method only to provide fake data. However, even using the make method, apparently the related factories are using the create method.
My code has several relationships in the factories, but below is a simple example of the problem.

### Steps To Reproduce:
PositionFactory:
```php
<?php

namespace Database\Factories;

use Illuminate\Database\Eloquent\Factories\Factory;
use App\Models\Firm;
use App\Models\Position;

class PositionFactory extends Factory
{
    /**
     * @var string
     */
    protected $model = Position::class;

    /**
     * @return array
     */
    public function definition(): array
    {
        return [
            'name' => $this->faker->name,
            'description' => $this->faker->text,
            'firm_id' => Firm::factory(),
        ];
    }
}

``` 
FirmFactory:
```php
<?php

namespace Database\Factories;

use Illuminate\Database\Eloquent\Factories\Factory;
use App\Models\Firm;

class FirmFactory extends Factory
{
    /**
     * @var string
     */
    protected $model = Firm::class;

    /**
     * @return array
     */
    public function definition(): array
    {
        return [
            'name' => $this->faker->company,
            'description' => $this->faker->paragraph,
            'website' => $this->faker->url,
            'discount' => $this->faker->randomFloat(),
            'supplier' => $this->faker->word,
            'status' => $this->faker->randomElement(['active', 'disabled']),
        ];
    }
}

``` 

The error presented is when running the tests, it is the attempt to create a user. I imagine it’s because most relationships are tied to that table
```shell
  • Tests\UT\Repositories\TemplateFieldPropertyValueRepositoryTest > attach dependency existing dependency
   Illuminate\Database\QueryException 

  SQLSTATE[HY000]: General error: 1 no such table: users (SQL: insert into "users" ("first_name", "last_name", "email", "password", "secret_phrase", "updated_at", "created_at") values (Fritz, Murray, tania32@stroman.org, $2y$10$Tc6AvPJ3Ijsh2kObc708NuCjscMlgOkrH8yxaBD9CBJZLSiF48pIC, Quis occaecati optio suscipit quia rerum esse. Sapiente voluptatem molestias temporibus repudiandae et. Illo at distinctio cum corporis placeat. Sed quam occaecati sed non., 2021-02-06 23:24:37, 2021-02-06 23:24:37))

  at vendor/illuminate/database/Connection.php:678
    674▕         // If an exception occurs when attempting to run a query, we'll format the error
    675▕         // message to include the bindings with SQL, which will make this exception a
    676▕         // lot more helpful to the developer instead of just the database's errors.
    677▕         catch (Exception $e) {
  ➜ 678▕             throw new QueryException(
    679▕                 $query, $this->prepareBindings($bindings), $e
    680▕             );
    681▕         }
    682▕ 

  1   [internal]:0
      Illuminate\Database\Eloquent\Factories\Factory::Illuminate\Database\Eloquent\Factories\{closure}(Object(Database\Factories\UserFactory), "user_id")

  2   [internal]:0
      Illuminate\Database\Eloquent\Factories\Factory::Illuminate\Database\Eloquent\Factories\{closure}(Object(Database\Factories\TemplateFactory), "template_id")


  Tests:  2 failed
  Time:   0.17s

```
Debugging the code to solve the problem, I found the place where the create method is called.
[Illuminate/Database/Eloquent/Factories/Factory.php](https://github.com/laravel/framework/blob/v8.22.0/src/Illuminate/Database/Eloquent/Factories/Factory.php#L413)
```php
    /**
     * Expand all attributes to their underlying values.
     *
     * @param  array  $definition
     * @return array
     */
    protected function expandAttributes(array $definition)
    {
        return collect($definition)->map(function ($attribute, $key) use (&$definition) {
            if (is_callable($attribute) && ! is_string($attribute) && ! is_array($attribute)) {
                $attribute = $attribute($definition);
            }

            if ($attribute instanceof self) {
                $attribute = $attribute->create()->getKey(); // <======= HERE
            } elseif ($attribute instanceof Model) {
                $attribute = $attribute->getKey();
            }

            $definition[$key] = $attribute;

            return $attribute;
        })->all();
    }
```
I believe this code should check first, whether it is called by the make or create method, and then use the correct method.
(If I switch to the make method, the tests work).

**Comment by robertoarruda** [comment]
@taylorotwell sorry, but in my opinion, this behavior is an error. Because as the doc says “the make method to create models without persisting them to the database”. Therefore, you should not use the database at any time.

I had to install an SQLite database in memory and use the migration to be able to use the factories in my unit tests. However, the performance of the tests was very poor. Before, tests took about 2 minutes with xdebug and coverage enabled. Now, without xdebug and coverage, it takes 16 minutes. Very bad :tired_face:

**Comment by adri3nam** [comment]
Exactly the same situation than @robertoarruda .
I'm coding an package to share models and factories to different Laravel projects. This package run some test and I don't need a DB for these tests. For factories without dependency there is no problem I can call make() but if there is one dependency defined inside the definition() method, this dependency will be saved in database and I don't expect this behavior with make(). I searched for a workaround but I can't find one. I also started to think about SQLite but I also need to maintain migrations with the packages and find how to run it without the entire Laravel framework (because it's only a pacjkage).

**Comment by robertoarruda** [comment]
@adri3nam To get around this problem (which still has no solution), I made this horrible code on my system.
I needed to override some methods from Laravel's original Factory class and use them in my Factory classes.

```php
<?php

namespace Database\Factories;

use Illuminate\Database\Eloquent\Collection;
use Illuminate\Database\Eloquent\Factories\Factory as BaseFactory;
use Illuminate\Database\Eloquent\Model;

abstract class Factory extends BaseFactory
{
    /**
     * @var bool
     */
    private bool $isCreating = false;

    /**
     * @param array $attributes
     * @param Model|null $parent
     * @return Collection|Model|mixed
     */
    public function create($attributes = [], ?Model $parent = null)
    {
        $this->isCreating = true;

        return parent::create($attributes, $parent);
    }

    /**
     * @param array $definition
     * @return array
     */
    protected function expandAttributes(array $definition)
    {
        return collect($definition)->map(function ($attribute, $key) use (&$definition) {
            if (is_callable($attribute) && ! is_string($attribute) && ! is_array($attribute)) {
                $attribute = $attribute($definition);
            }

            if ($attribute instanceof self) {
                $attribute = ($this->isCreating) ? $attribute->create()->getKey() : 0;
            } elseif ($attribute instanceof Model) {
                $attribute = $attribute->getKey();
            }

            $definition[$key] = $attribute;

            return $attribute;
        })->all();
    }
}
```

**Reply by adri3nam** [reply]
Thank you @robertoarruda !
I was thinking about something similar even if I don't really like the idea of rewriting expandAttributes() but I see no elegant way to do it differently.

**Reply by robertoarruda** [reply]
I agree with you, this shouldn't be the only solution to do this. So I recreated the issue so that this error is fixed.
https://github.com/laravel/framework/issues/41313

**Comment by ekateiva** [comment]
As the Laravel core is not willing to change the behaviour, an alternative without changing the `expandAttributes()` could be:

A trait similar to this:
```php
use Illuminate\Database\Eloquent\Factories\Factory;

trait WithoutRelationships
{
    public function withoutRelationships(): static
    {
        return $this->state(function () {
            return collect($this->definition())
                ->filter(function ($value) {
                    return $value instanceof Factory;
                })
                ->mapWithKeys(function ($item, $key) {
                    return [$key => null];
                })
                ->toArray();
        });
    }
}
```

Include this trait in your model Factory:
```php
class SomeFactory extends Factory
{
    use WithoutRelationships;

    ...
}
```

And then you could use this code in your tests:
```php
SomeModel::factory()->withoutRelationships()->make();
```

This would allow you to not care if you add another Model::factory() as a property later on.

**Reply by robertoarruda** [reply]
You can make your suggestion in this issue https://github.com/laravel/framework/issues/41313
The suggestions made there were not accepted.
